The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3405266244/jobs/5663118134) [(plain)](https://github.com/rust-lang-ci/rust/commit/3d331029b8b114cffdc374f6166ca072a914f4e8/checks/9322467476/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_symbol_mangling test:false 8.992
error[E0308]: mismatched types
   --> compiler/rustc_trait_selection/src/traits/project.rs:654:24
    |
648 |     fn fold_const(&mut self, constant: ty::Const<'tcx>) -> ty::Const<'tcx> {
    |                                                            --------------- expected `rustc_middle::ty::Const<'tcx>` because of return type
654 |                 return Ok(constant);
    |                        ^^^^^^^^^^^^ expected struct `rustc_middle::ty::Const`, found enum `Result`
    |
    = note: expected struct `rustc_middle::ty::Const<'_>`
```

</details>

--- a/compiler/rustc_trait_selection/src/traits/project.rs
+++ b/compiler/rustc_trait_selection/src/traits/project.rs
@@ -647,7 +647,7 @@ impl<'a, 'b, 'tcx> TypeFolder<'tcx> for AssocTypeNormalizer<'a, 'b, 'tcx> {
     #[instrument(skip(self), level = "debug")]
     fn fold_const(&mut self, constant: ty::Const<'tcx>) -> ty::Const<'tcx> {
         let tcx = self.selcx.tcx();
-        if tcx.lazy_normalization() {
+        if tcx.lazy_normalization() || !needs_normalization(&constant, self.param_env.reveal()) {
             constant
         } else {
             let constant = constant.super_fold_with(self);

--- a/compiler/rustc_trait_selection/src/traits/query/normalize.rs
+++ b/compiler/rustc_trait_selection/src/traits/query/normalize.rs
@@ -353,6 +353,10 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
         &mut self,
         constant: ty::Const<'tcx>,
     ) -> Result<ty::Const<'tcx>, Self::Error> {
+        if !needs_normalization(&constant, self.param_env.reveal()) {
+            return Ok(constant);
+        }
+
         let constant = constant.try_super_fold_with(self)?;
         debug!(?constant, ?self.param_env);
         Ok(crate::traits::project::with_replaced_escaping_bound_vars(

